lib/skate/settings/user.ex

```diff
@@ -2,16 +2,61 @@ defmodule Skate.Settings.User do
   alias Skate.Settings.Db.User, as: DbUser
 
   import Ecto.Query
+  require Logger
+
+  @spec get(String.t()) :: DbUser.t()
+  @doc """
+  Get a user by their username
+  """
+  def get(username) do
+    Skate.Repo.get_by!(DbUser, username: username)
+  end
+
+  @spec get_by_email(String.t()) :: DbUser.t() | nil
+  @doc """
+  Get a user with the matching email, if one exists
+  """
+  def get_by_email(email) do
+    Skate.Repo.get_by(DbUser, email: String.downcase(email))
+  end
+
+  @spec upsert(username :: String.t(), email :: String.t()) :: DbUser.t()
+  @doc """
+  Update the user with the given username if one exists, otherwise insert a new one.
+  """
+  def upsert(username, email) do
+    email = String.downcase(email)
+    user_matching_email = get_by_email(email)
 
-  def get_or_create(username) do
     user =
-      Skate.Repo.insert!(
-        DbUser.changeset(%DbUser{}, %{username: username}),
-        returning: true,
-        conflict_target: [:username],
-        # update a row with no effect so the returning works
-        on_conflict: {:replace, [:username]}
-      )
+      cond do
+        is_nil(user_matching_email) ->
+          Skate.Repo.insert!(
+            DbUser.changeset(%DbUser{}, %{
+              username: username,
+              email: email
+            }),
+            returning: true,
+            conflict_target: [:username],
+            on_conflict: {:replace, [:email]}
+          )
+
+        # existing user has same username - no update needed
+        user_matching_email.username == username ->
+          user_matching_email
+
+        # username format has changed. Create a new record for this user without an associated email
+        # This record will be used only temporarily until user settings are universally fetched by email.
```

Comment:
Handling of this case based on [this thread](https://github.com/mbta/skate/pull/1740#discussion_r981635623)